source/src/BNFC/Backend/Haskell/CFtoHappy.hs

```diff
@@ -133,17 +141,31 @@ rulesForHappy absM functor cf = for (ruleGroups cf) $ \ (cat, rules) ->
 -- Coercion are much simpler:
 --
 -- >>> constructRule "Foo" True (npRule "_" (Cat "Exp") [Right "(", Left (Cat "Exp"), Right ")"] Parsable)
--- ("'(' Exp ')'","$2")
+-- ("'(' Exp ')'","(Just (tokenLineCol $1), (snd $2))")
 --
 constructRule :: IsFun f => String -> Bool -> Rul f -> (Pattern, Action)
 constructRule absName functor (Rule fun0 _cat rhs Parsable) = (pattern, action)
   where
     fun = funName fun0
-    (pattern, metavars) = generatePatterns rhs
-    action | isCoercion fun                 = unwords metavars
-           | isNilCons fun                  = unwords (qualify fun : metavars)
-           | functor                        = unwords (qualify fun : "()" : metavars)
+    (pattern, metavars) = generatePatterns rhs functor
+    action | isCoercion fun && not functor  = unwords metavars
+           | isNilCons  fun && not functor  = unwords (qualify fun : metavars)
+           | functor                        = getFunctorAction
            | otherwise                      = unwords (qualify fun : metavars)
+    -- Commelina, 2020-12-10:
+    -- replace the previous unimplemented "()"
+    getFunctorPos rhs = case rhs of
+      []          -> "Nothing"
+      (Left _:_)  -> "fst $1"
+      (Right _:_) -> "Just (tokenLineCol $1)"
+    getFunctorValue
```

Comment:
Maybe it refers to `getFunctorValue`...? But indeed, the code needs refactoring.

After refactoring, the hunk contains three core functions `actionPos`, `actionValue` and `action`. And now `actionValue` can handle both situations with/without functor so there is less duplication. `actionPos` gets position information and `action` just composes them in different forms depends on `functor`:

```haskell
    action
      | functor   = let pos = actionPos
                        val = actionValue
                     in "(" ++ pos ++ ", " ++ val ++ ")"
      | otherwise = actionValue
    actionPos = case rhs of
      []          -> "Nothing"
      (Left _:_)  -> "fst $1"
      (Right _:_) -> "Just (tokenLineCol $1)"
    actionValue
      | isCoercion fun = unwords metavars
      | isNilCons  fun = unwords (qualify fun : metavars)
      | functor        = unwords (qualify fun : ("(" ++ actionPos ++ ")") : metavars)
      | otherwise      = unwords (qualify fun : metavars)
```